Framework/CurveFitting/src/Algorithms/SplineBackground.cpp

```diff
@@ -39,142 +39,263 @@ void SplineBackground::init() {
+  allocateBSplinePointers(numBins, ncoeffs);
+  addWsDataToSpline(inWS.get(), spec, numBins);
 
-  gsl_vector *c, *w, *x, *y;
-  gsl_matrix *Z, *cov;
-  gsl_multifit_linear_workspace *mw;
-  double chisq;
+  const auto &xInputPoints = inWS->x(spec);
+  setupSpline(xInputPoints.front(), xInputPoints.back(), numBins, ncoeffs);
 
-  int n = static_cast<int>(yInputVals.size());
-  bool isMasked = inWS->hasMaskedBins(spec);
-  std::vector<int> masked(yInputVals.size());
-  if (isMasked) {
-    auto maskedBins = inWS->maskedBins(spec);
-    for (auto &maskedBin : maskedBins)
-      masked[maskedBin.first] = 1;
-    n -= static_cast<int>(inWS->maskedBins(spec).size());
+  // Wrap this in its own block so we can alias the member variable
+  // to a shorter name for the duration of this call
+  {
+    // Have to pass a pointer so it can write chi ^ 2 even though we don't
+    // use this value
+    double chisq;
+    // Create temporary alias to reduce the size of the call
+    bSplinePointers &sp = m_splinePointers;
+    // do the fit
+    gsl_multifit_wlinear(sp.fittedWs, sp.binWeights, sp.yData, sp.coefficients,
+                         sp.covariance, &chisq, sp.weightedLinearFitWs);
   }
 
-  if (n < ncoeffs) {
-    g_log.error("Too many basis functions (NCoeff)");
-    throw std::out_of_range("Too many basis functions (NCoeff)");
-  }
+  auto outWS = saveSplineOutput(inWS, spec);
+  freeBSplinePointers();
+  setProperty("OutputWorkspace", outWS);
+}
 
-  /* allocate a cubic bspline workspace (k = 4) */
-  bw = gsl_bspline_alloc(k, nbreak);
-  B = gsl_vector_alloc(ncoeffs);
+/**
+ * Calculates the number of unmasked bins to process within spline background
+ *
+ * @param ws:: The input workspace to calculate the number of unmasked bins of
+*/
+size_t
+SplineBackground::calculateNumBinsToProcess(const API::MatrixWorkspace *ws) {
+  const int spec = getProperty("WorkspaceIndex");
+  const auto &yInputVals = ws->y(spec);
+  size_t ySize = yInputVals.size();
+
+  // Check and subtract masked bins from the count
+  bool isMasked = ws->hasMaskedBins(spec);
+  if (isMasked) {
+    ySize -= static_cast<int>(ws->maskedBins(spec).size());
+  }
+  return ySize;
+}
 
-  x = gsl_vector_alloc(n);
-  y = gsl_vector_alloc(n);
-  Z = gsl_matrix_alloc(n, ncoeffs);
-  c = gsl_vector_alloc(ncoeffs);
-  w = gsl_vector_alloc(n);
-  cov = gsl_matrix_alloc(ncoeffs, ncoeffs);
-  mw = gsl_multifit_linear_alloc(n, ncoeffs);
+/**
+ * Adds the data from the workspace into various GSL vectors
+ * for the spline later
+ *
+ * @param ws:: The workspace containing data to process
+ * @param specNum:: The spectrum number to process from the input WS
+ * @param expectedNumBins:: The expected number of unmasked bins to add
+ */
+void SplineBackground::addWsDataToSpline(const API::MatrixWorkspace *ws,
+                                         const size_t specNum,
+                                         int expectedNumBins) {
+  const auto xInputVals = ws->points(specNum);
```

Comment:
Similar to my previous comment, in this case you are supplying points rather than bin edges, so I suggest changing the name to xInputPoints.